libflux/src/semantic/types.rs

```diff
@@ -14,57 +19,138 @@ use std::{cmp, collections::HashMap, fmt};
 #[derive(Debug, Clone, PartialEq)]
 pub struct PolyType {
     pub free: Vec<Tvar>,
-    pub bnds: Option<HashMap<Tvar, Kind>>,
+    pub cons: Option<Const>,
     pub expr: MonoType,
 }
 
 impl fmt::Display for PolyType {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match &self.bnds {
-            Some(bnds) => {
-                let mut bounds = Vec::new();
-                for tv in &self.free {
-                    if let Some(kind) = bnds.get(tv) {
-                        bounds.push(BoundTvar {
-                            tv: *tv,
-                            kind: *kind,
-                        })
-                    }
-                }
-                write!(
-                    f,
-                    "forall [{}] where {} {}",
-                    DisplayList {
-                        values: &self.free,
-                        delim: ", "
-                    },
-                    DisplayList {
-                        values: &bounds,
-                        delim: ", "
-                    },
-                    self.expr,
-                )
-            }
-            None => write!(
+        let vars = self.free.iter().map(|x| x.to_string()).collect::<Vec<_>>();
+        match &self.cons {
+            Some(constraints) => write!(
                 f,
-                "forall [{}] {}",
-                DisplayList {
-                    values: &self.free,
-                    delim: ", "
-                },
-                self.expr
+                "forall [{}] where {} {}",
+                vars.join(", "),
+                constraints,
+                self.expr,
+            ),
+            None => write!(f, "forall [{}] {}", vars.join(", "), self.expr),
+        }
+    }
+}
+
+// Const is used to keep track of the generic constraints for each type
+// variable during type inference.
+//
+// Note that during inference we might infer that a type variable is of
+// a particular kind (type class) without inferring an exact monotype
+// for said type variable. This information is stored in the Const data
+// structure.
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct Const(HashMap<Tvar, HashSet<Kind>>);
```

Comment:
Why not `Constraints`? `Const` makes me think about "constant".

Reply:
There's already a `Constraints` struct for the constraints that inference generates. Maybe `TvarKinds`?